aeneas/src/ir/PackingSolver.v3

```diff
@@ -0,0 +1,332 @@
+
+Based on each class, the packing solver is able to squeeze other fields into the
+unassigned portions of the representation.
+*/
+
+type Bit {
+	case Unassigned; // ?
+	case Fixed(v: u1); // 0 or 1
+	case Unusable; // .
+}
+
+ // TODO: Allow non-contiguous intervals to be represented
+type Interval(s: int, e: int) {
+	def render(sb: StringBuilder) -> StringBuilder {
+		return sb.put2("%d-%d", s, e);
+	}
+}
+
+component PackingStates {
+	def fromString(s: string) -> PackingState {
+		var bits = Vector<Bit>.new();
+		for (c in s) {
+			match (c) {
+				_ => ;
+				'0' => bits.put(Bit.Fixed(0));
+				'1' => bits.put(Bit.Fixed(1));
+				'.' => bits.put(Bit.Unusable);
+				'?' => bits.put(Bit.Unassigned);
+			}
+		}
+		for (i < bits.length / 2) {
+			var x = bits[i];
+			bits[i] = bits[bits.length - i - 1];
+			bits[bits.length - i - 1] = x;
+		}
+		return PackingState.new(bits.extract());
+	}
+
+	def mergeBit(a: Bit, b: Bit) -> (Bit, bool) {
+		match (a) {
+			Unassigned => return (b, true);
+			Unusable => match (b) {
+				Unassigned => return (Bit.Unusable, true);
+				_ => return (Bit.Unusable, false);
+			}
+			Fixed(x) => match (b) {
+				Unusable => return (a, false);
+				Unassigned => return (a, true);
+				Fixed(y) => return (a, x == y);
+			}
+		}
+	}
+	def mergeBits(a: Array<Bit>, b: Array<Bit>) -> Array<Bit> {
+		if (a.length != b.length) return null;
+		var c = Array<Bit>.new(a.length);
+		for (i < a.length) {
+			var ab = mergeBit(a[i], b[i]);
+			if (!ab.1) return null;
+			c[i] = ab.0;
+		}
+		return c;
+	}
+}
+
+class PackingState(bits: Array<Bit>) {
+	def size = bits.length;
+	
+	def render(s: StringBuilder) -> StringBuilder {
+		for (i < bits.length) {
+			var b = bits[bits.length - i - 1];
+			if (i != 0 && i % 8 == 0) s.putc('_');
+			match (b) {
+				Unassigned => s.putc('?');
+				Fixed(v) => s.put1("%d", v);
+				Unusable => {
+					// s.puts(Terminal.CTRL_GREEN);
+					// s.pututf8(0x25ae);
+					// s.puts(Terminal.CTRL_DEFAULT);
+					s.putc('.');
+				}
+			}
+		}
+		return s;
+	}
+	def getIntervalsForSize(s: int) -> Array<Interval> {
+		var i = 0;
+		var intervals = Vector<Interval>.new();
+
+		for (i < size - s + 1) {
+			var start = size - s - i;
+			var isEmpty = true;
+			for (j < s) {
+				if (bits[start + j] != Bit.Unassigned) {
+					isEmpty = false;
+					break;
+				}
+			}
+			if (isEmpty) intervals.put(Interval(start, start + s));
+		}
+		return intervals.extract();
+	}
+	def assignInterval(i: Interval) {
+		for (j = i.s; j < i.e; j++) bits[j] = Bit.Unusable;
+	}
+	def unassignInterval(i: Interval) {
+		for (j = i.s; j < i.e; j++) bits[j] = Bit.Unassigned;
+	}
+	def copy() -> PackingState { return PackingState.new(Arrays.dup(bits)); }
+}
+
+type PackingField {
+	case Nonref(size: int);
+	case Ref;
+}
+
+type FieldLoc(caseIdx: int, fieldIdx: int) {
+	def hash() -> int { return caseIdx * 31 + fieldIdx; }
+
+	def equal(other: FieldLoc) -> bool { return caseIdx == other.caseIdx && fieldIdx == other.fieldIdx; }
```

Comment:
Because this is a datatype, the builtin `==` will do a structural check, so this method isn't needed.